### PR TITLE
Rank same kanji higher than different kanji with the same reading#67

### PR DESCRIPTION
**Pull request by prazuma**
- if there are records record1:[name=菊地真], record2:[name=菊地眞], get results(1st: 菊地真、2nd: 菊地眞) when search by "菊地 真"
- demo: http://pfrank-1078.appspot.com/global/home.html

### CODE REVIEW

File: app/full_text_search.py [hunk 1]
```diff
@@ -213,6 +242,12 @@ def create_romanized_name_fields(romanize_method, **kwargs):
         if romanized_name:
             fields.extend(create_fields_for_rank(field, romanized_name))
             romanized_names_list.append(romanized_name)
+        if script_variant.has_kanji(kwargs[field]):
```
Comment (gimite) [inline app/full_text_search.py]: You may not limit to kanji. You may always add this _non_romanized field and rank exact matches higher than others. What do you think?

File: app/full_text_search.py [hunk 2]
```diff
@@ -109,12 +113,35 @@ def search(repo, query_txt, max_results):
...
+
+    person_location_index_results_with_kanji = person_location_index.search(
+        appengine_search.Query(
+            query_string=and_query_with_kanji, options=options)
+    )
+
     index_results = []
-    regexp = make_or_regexp(query_txt)
+    regexp = make_or_regexp(romanized_query)
+    for document in person_location_index_results_with_kanji:
```
Comment (gimite) [inline app/full_text_search.py]: The code here is very similar to the code at line 145. Can you define a function instead?

In general, repeating the same code multiple times is considered bad. It makes harder to read, and you need to update both codes when you need to modify the code. Please consider defining a function instead before you copy&paste multiple lines of code.

File: app/script_variant.py [hunk 3]
```diff
@@ -35,6 +35,8 @@ def has_kanji(word):
     """
     # [\u3400-\u9fff]: kanji
     # TODO: update this range (some of kanji characters out of this range.)
+    if not word:
```
Comment (gimite) [inline app/script_variant.py]: Do you need this code? I believe it anyway just returns False if word is empty.

File: tests/test_full_text_search.py [hunk 4]
```diff
@@ -254,6 +269,12 @@ def test_search_by_name_only(self):
         assert set([r.record_id for r in results]) == \
             set(['haiti/0225'])
 
+        # Check rank order
+        # (same kanji higher than different kanji with the same reading)
+        results = full_text_search.search('haiti', u'菊地 真', 5)
```
Comment (gimite) [inline tests/test_full_text_search.py]: Add test with [菊池 眞] too. If we test with just one of them, there is possibility that ['haiti/0829', 'haiti/1829'] are always ranked in this order regardless of the query.

File: app/full_text_search.py (hunk 2 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 1 above)
Comment (prazuma) [inline app/full_text_search.py]: fixed

File: app/script_variant.py (hunk 3 above)
Comment (prazuma) [inline app/script_variant.py]: done

File: tests/test_full_text_search.py (hunk 4 above)
Comment (prazuma) [inline tests/test_full_text_search.py]: done

File: app/full_text_search.py [hunk 5]
```diff
@@ -54,7 +54,7 @@ def create_sort_expressions():
     return [appengine_search.SortExpression(
         expression='_score',
         direction=appengine_search.SortExpression.DESCENDING,
-        default_value=0.0
+        default_value=1.0
```
Comment (gimite) [inline app/full_text_search.py]: Why did you change this to 1.0?

File: app/full_text_search.py [hunk 6]
```diff
@@ -244,9 +267,10 @@ def create_romanized_name_fields(romanize_method, **kwargs):
     for field_name in kwargs:
         romanized_names = romanize_method(kwargs[field_name])
         for index, romanized_name in enumerate(romanized_names):
-            fields.extend(create_fields_for_rank('%s_%d' %
-                                                 (field_name, index),
+            fields.extend(create_fields_for_rank('%s_%s_%d' %
+                                                 (field_name, romanize_method_name, index),
```
Comment (gimite) [inline app/full_text_search.py]: Over 80 chars.

File: app/full_text_search.py [hunk 7]
```diff
@@ -244,9 +267,10 @@ def create_romanized_name_fields(romanize_method, **kwargs):
     for field_name in kwargs:
         romanized_names = romanize_method(kwargs[field_name])
         for index, romanized_name in enumerate(romanized_names):
-            fields.extend(create_fields_for_rank('%s_%d' %
-                                                 (field_name, index),
+            fields.extend(create_fields_for_rank('%s_%s_%d' %
```
Comment (gimite) [inline app/full_text_search.py]: Add a comment why you add both of romanized and non-romanized fields.

File: app/full_text_search.py [hunk 8]
```diff
@@ -244,9 +267,10 @@ def create_romanized_name_fields(romanize_method, **kwargs):
...
+                                                 (field_name, romanize_method_name, index),
                                                  romanized_name))
+            fields.extend(create_fields_for_rank('%s_%d' % (field_name, index), kwargs[field_name]))
```
Comment (gimite) [inline app/full_text_search.py]: Do you need to add this field (with the same value) len(romanized_names) times?

File: app/full_text_search.py [hunk 9]
```diff
@@ -244,9 +267,10 @@ def create_romanized_name_fields(romanize_method, **kwargs):
     for field_name in kwargs:
```
Comment (gimite) [inline app/full_text_search.py]: You can write

    for field_name, field_value in kwargs.iteritems():

to get the value instead.

File: app/full_text_search.py [hunk 10]
```diff
@@ -125,6 +150,11 @@ def search(repo, query_txt, max_results):
     # Remove double quotes so that we can safely apply enclose_in_double_quotes().
     query_txt = re.sub('"', '', query_txt)
     romanized_query = create_query_txt(query_txt)
+    kanji_words_in_query_txt = script_variant.find_kanji_word(query_txt)
```
Comment (gimite) [inline app/full_text_search.py]: If you want to apply the logic for non-kanji words as well, you should extract all words here instead of kanji words.

File: app/full_text_search.py [hunk 11]
```diff
@@ -101,6 +101,31 @@ def create_query_txt(query_txt):
     return enclose_in_parenthesis(romanized_query)
 
 
+def get_person_ids_from_results(romanized_query, results_list):
+    """
+    Returns person record_id (a part of romanized_query matches person name) list.
```
Comment (gimite) [inline app/full_text_search.py]: "Returns record_id of persons whose name contains at least one word in romanized_query."

Also, please mention that:

- Why you need to apply this regexp
- It also removes dups (i.e., If results_list contains multiple results with the same index_results, it returns just one of them.)

File: app/full_text_search.py [hunk 12]
```diff
@@ -125,6 +150,11 @@ def search(repo, query_txt, max_results):
     # Remove double quotes so that we can safely apply enclose_in_double_quotes().
     query_txt = re.sub('"', '', query_txt)
     romanized_query = create_query_txt(query_txt)
+    kanji_words_in_query_txt = script_variant.find_kanji_word(query_txt)
+    double_quote_kanji_words_in_query_txt = ' '.join(
+        enclose_in_double_quotes(word) for word in kanji_words_in_query_txt)
+    romanized_query_with_kanji = romanized_query + ' ' +\
```
Comment (gimite) [inline app/full_text_search.py]: Do you need to add romanized_query? Would it work if you search just with double_quote_kanji_words_in_query_txt?

File: app/full_text_search.py [hunk 13]
```diff
@@ -144,27 +174,20 @@ def search(repo, query_txt, max_results):
...
-            if field.name == 'names_romanized_by_romanize_word_by_unidecode':
-                names = field.value
-            if field.name == 'record_id':
-                id = field.value
-            if field.name == 'names_romanized_by_romanize_japanese_name_by_name_dict':
-                romanized_jp_names = field.value
-
-        if regexp.search(names) or regexp.search(romanized_jp_names):
-            index_results.append(id)
+    person_location_index_results_with_kanji = person_location_index.search(
```
Comment (gimite) [inline app/full_text_search.py]: Add a comment why you need to perform this search in addition to person_location_index_results above.

File: app/full_text_search.py (hunk 5 above)
Comment (prazuma) [inline app/full_text_search.py]: fixed

File: app/full_text_search.py (hunk 11 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 12 above)
Comment (prazuma) [inline app/full_text_search.py]: removed romanized_query

File: app/full_text_search.py (hunk 13 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 9 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 7 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 6 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 8 above)
Comment (prazuma) [inline app/full_text_search.py]: fixed

File: app/full_text_search.py [hunk 14]
```diff
@@ -281,6 +317,17 @@ def create_romanized_location_fields(romanize_method, **kwargs):
             )
     return fields
 
+def create_non_romanized_fields(**kwargs):
+    """
+    Creates non romanized fields for ranking same kanji higher than
```
Comment (gimite) [inline app/full_text_search.py]: Now this is not limited to kanji, so you should instead say e.g., "to rank exact matches higher than non-exact matches with the same romanization". Also it would be nice to add example here.

File: app/full_text_search.py [hunk 15]
```diff
@@ -101,6 +101,37 @@ def create_query_txt(query_txt):
     return enclose_in_parenthesis(romanized_query)
 
 
+def get_person_ids_from_results(romanized_query, results_list):
+    """
+    Returns person record_id of persons
+    whose name contains at least one word in romanized_query.
+    We need regexp because check romanized_query matches
```
Comment (gimite) [inline app/full_text_search.py]: Please fix this comment as discussed offline.

File: app/full_text_search.py (hunk 12 above)
Comment (gimite) [inline app/full_text_search.py]: It is not removed yet.

File: app/full_text_search.py [hunk 16]
```diff
@@ -149,22 +186,19 @@ def search(repo, query_txt, max_results):
...
-            if field.name == 'names_romanized_by_romanize_word_by_unidecode':
-                names = field.value
-            if field.name == 'record_id':
-                id = field.value
-            if field.name == 'names_romanized_by_romanize_japanese_name_by_name_dict':
-                romanized_jp_names = field.value
-
-        if regexp.search(names) or regexp.search(romanized_jp_names):
-            index_results.append(id)
+    # For ranking same kanji higher than different kanji with the same reading.
```
Comment (gimite) [inline app/full_text_search.py]: Now this is not limited to kanji, so you should instead say e.g., "To rank exact matches higher than non-exact matches with the same romanization".

File: app/full_text_search.py (hunk 10 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 15 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 14 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 16 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 12 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py [hunk 17]
```diff
@@ -101,6 +114,38 @@ def create_query_txt(query_txt):
     return enclose_in_parenthesis(romanized_query)
 
 
+def get_person_ids_from_results(romanized_query, results_list):
+    """
+    Returns person record_id of persons
+    whose name contains at least one word in romanized_query.
+    We need regexp to check romanized_query matches
+    at least a part of person name.
+    To protect users' privacy, we should not return records matches location only.
```
Comment (gimite) [inline app/full_text_search.py]: Please fix English grammar here as chatted offline.

File: app/full_text_search.py [hunk 18]
```diff
@@ -144,27 +190,21 @@ def search(repo, query_txt, max_results):
...
-            if field.name == 'record_id':
-                id = field.value
-            if field.name == 'names_romanized_by_romanize_japanese_name_by_name_dict':
-                romanized_jp_names = field.value
-
-        if regexp.search(names) or regexp.search(romanized_jp_names):
-            index_results.append(id)
+    # To rank exact matches higher than non-exact matches with the same romanization.
+    non_romanized_and_query = non_romanized_query + ' AND (repo: ' + repo + ')'
+    person_location_index_results_with_kanji = person_location_index.search(
```
Comment (gimite) [inline app/full_text_search.py]: Since it's no longer limited to kanji, please rename this to e.g., "non_romanized_person_location_index_results".

File: app/full_text_search.py [hunk 19]
```diff
@@ -241,15 +281,17 @@ def create_romanized_name_fields(romanize_method, **kwargs):
...
-            fields.extend(create_fields_for_rank('%s_%d' %
-                                                 (field_name, index),
+            fields.extend(create_fields_for_rank('%s_%s_%d' %
+                                                 (field_name,
+                                                  romanize_method_name, index),
                                                  romanized_name))
         romanized_names_list.extend(romanized_names)
 
-    full_name_fields, romanized_full_names = create_full_name_without_space_fields(
+    full_name_fields, romanized_full_names =\
```
Comment (gimite) [inline app/full_text_search.py]: Insert white space between "=" and "\". Or put "(" instead here and close it at the end of the statement.

File: app/full_text_search.py (hunk 17 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 18 above)
Comment (prazuma) [inline app/full_text_search.py]: done

File: app/full_text_search.py (hunk 19 above)
Comment (prazuma) [inline app/full_text_search.py]: done

**Comment by gimite** [comment]
LGTM. Please merge it after running tools/all_tests.
